src/doc/rustc/src/platform-support/bpf-unknown-none.md

```diff
@@ -0,0 +1,157 @@
+```
+
+## Building Rust programs
+
+Rust does not yet ship pre-compiled artifacts for this target. To compile for
+this target, you will either need to build Rust with the target enabled (see
+"Building the target" above), or build your own copy of `core` by using
+`build-std` or similar.
+
+Building the BPF target requires specifying it explicitly. Users can either
+add it to the `target` list in `config.toml`:
+
+```toml
+[build]
+target = ["bpfel-unknown-none"]
+```
+
+Or specify it directly in the `cargo build` invocation:
+
+```console
+cargo +nightly build -Z build-std=core --target bpfel-unknown-none
+```
+
+BPF has its own debug info format called [BTF][btf]. The BPF backend in LLVM
+simply converts the LLVM debug info into BTF.
+
+BPF targets use [bpf-linker][bpf-linker], an LLVM bitcode linker, which by
+default strips the debug info, but it has an experimental feature of emitting
+BTF, which can be enabled by adding `-C link-arg=--btf` to `RUSTFLAGS`. With
+that feature enabled, [bpf-linker][bpf-linker] does not only link different
+crates/modules, but also performs a necessary sanitization of debug info, which
+is required to produce valid [BTF][btf] acceptable by the Linux kernel.
+
+## Error handling
+
+There is no concept of stack unwinding in BPF, therefore BPF programs are
+expected to handle errors in a recoverable manner. Therefore most BPF programs
+written in Rust use the following no-op panic handler implementation:
+
+```rust
+#[cfg(not(test))]
+#[panic_handler]
+fn panic(_info: &core::panic::PanicInfo) -> ! {
+    loop {}
+}
+```
+
+Infinite loops are forbidden by the BPF verifier. Therefore, if the program
+contains any code which can panic, the BPF VM rejects to load it.
+
+## Testing
+
+BPF bytecode needs to be executed on a BPF virtual machine, like the one
+provided by the Linux kernel or one of the user-space implementations like
+[rbpf][rbpf]. None of them support running Rust testsuite. One of the reasons
+is the lack of support for panicking.
+
+Currently the best way of unit testing the code used in BPF programs is
+extracting it to a separate crate, which can be built and ran on a host target,
+and using it as a dependency.
+
+## Cross-compilation toolchains
+
+BPF programs are always cross-compiled from a host (e.g.
+`x86_64-unknown-linux-*`) for a BPF target (e.g. `bpfel-unknown-none`).
+
+The endianness of a chosen BPF target needs to match the endianness of the BPF
+VM host on which the program is supposed to run.
+
+The architecture of the BPF VM host often has an impact on types that the BPF
+programs should use. For example [kprobes][kprobe], [fprobes][fprobe] and
+[uprobes][uprobe] allow dynamic function tracing and lookup into registers.
+Registers and calling convenctions are different across architectures. Due to
+`char` in C being signed or unsigned on different architectures, the signatures
+of [BPF helpers][bpf-helpers] and [kfuncs][kfunc] also end up different.
```

Comment:
Doesn't Linux compile with `-funsigned-char` nowadays?

Reply:
From relatively recently right? 6.1 or 6.2 can't remember. But we need to support older kernels.

But I agree that it seems a detail I wouldn't single out `char` here.